lib/user_lookup.rb

```diff
@@ -12,66 +19,46 @@ def [](user_id)
   end
 
   def primary_groups
-    @primary_groups ||= begin
-      hash = {}
-      users.values.each do |u|
-        if u.primary_group_id
-          hash[u.id] = groups[u.primary_group_id]
-        end
+    @primary_groups ||= users
+      .values
+      .filter(&:primary_group_id)
+      .each_with_object({}) do |user, hash|
+        hash[user.id] = groups[user.primary_group_id]
       end
-      hash
-    end
   end
 
   def flair_groups
-    @flair_groups ||= begin
-      hash = {}
-      users.values.each do |u|
-        if u.flair_group_id
-          hash[u.id] = groups[u.flair_group_id]
-        end
+    @flair_groups ||= users
+      .values
+      .filter(&:flair_group_id)
+      .each_with_object({}) do |user, hash|
+        hash[user.id] = groups[user.flair_group_id]
       end
-      hash
-    end
   end
 
   private
 
-  def self.lookup_columns
-    @user_lookup_columns ||= %i{id username name uploaded_avatar_id primary_group_id flair_group_id admin moderator trust_level}
-  end
-
-  def self.group_lookup_columns
-    @group_lookup_columns ||= %i{id name flair_icon flair_upload_id flair_bg_color flair_color}
-  end
-
   def users
-    @users ||= user_lookup_hash
-  end
-
-  def user_lookup_hash
-    hash = {}
-    User.where(id: @user_ids)
+    @users ||= User
+      .where(id: @user_ids)
       .select(self.class.lookup_columns)
-      .each { |user| hash[user.id] = user }
-    hash
+      .index_by(&:id)
   end
 
   def groups
-    @group_lookup = begin
-      group_ids = users.values.map { |u| [u.primary_group_id, u.flair_group_id] }
-      group_ids.flatten!
-      group_ids.uniq!
-      group_ids.compact!
+    @group_lookup ||= begin
+      group_ids = users
+        .values
+        .map { |u| [u.primary_group_id, u.flair_group_id] }
+        .flatten
+        .uniq
+        .compact
```

Comment:
I prefer the previous implementation actually since it mutates the existing array instead of allocating new ones.

Reply:
As in the previous comment, `#flatten!` does create a new array under the hood, so does `#values` and `#map`. This doesn't change much.